File: examples/nodejs/whale.js
import { Config, TradeContext, HttpClient, TopicType } from 'longportwhale';
import pino from 'pino';

const logger = pino();

const conf = Config.fromEnv();
const httpcli = HttpClient.fromEnv();
const test_account = "L6VQEU00121996";

let orderId = null;
let completed = false;

function onOrder(err, ev) {
    if (err) {
        logger.error(err);
        return;
    }

    // Use toString() for logging (recommended workaround)
    logger.info(`Received order event: ${ev.toString()}`);

    // Access properties directly (recommended for production)
    const eventOrderId = ev.orderId;
    const eventSymbol = ev.symbol;
    const eventStatus = ev.status;

    logger.info(`Order details - ID: ${eventOrderId}, Symbol: ${eventSymbol}, Status: ${eventStatus}`);

    // Check if it matches our order
    if (orderId && eventOrderId === orderId) {
        logger.info("✅ Received expected order event!");
        completed = true;
    }
}

async function main() {
    logger.info("Starting verification...");

    const tctx = await TradeContext.new(conf);
    tctx.setOnOrderChanged(onOrder);

    logger.info("Subscribing to private events...");
    await tctx.subscribe([TopicType.Private]);

    // Wait for subscription
    await new Promise(resolve => setTimeout(resolve, 2000));

    // Submit Order
    const req = {
        "symbol": "700.HK",
        "order_type": "MO",
        "side": "Buy",
        "submitted_quantity": "100",
        "time_in_force": "Day",
        "account_no": test_account
    };

    logger.info(`Submitting order: ${JSON.stringify(req)}`);
    try {
        const submitRes = await httpcli.request('POST', '/v1/whaleapi/trade/order', null, req);
        logger.info(`Submit order response: ${JSON.stringify(submitRes)}`);
        orderId = submitRes.order_id;
    } catch (e) {
        logger.error(`Failed to submit order: ${e}`);
        process.exit(1);
    }

    // Wait for event
    logger.info(`Waiting for order event for order_id: ${orderId}...`);
    const startTime = Date.now();
    while (!completed) {
        if (Date.now() - startTime > 60000) {
            logger.error("Timeout waiting for order event.");
            process.exit(1);
        }
        await new Promise(resolve => setTimeout(resolve, 1000));
    }

    logger.info("Verification successful!");
    process.exit(0);
}

main().catch(e => {
    logger.error(e);
    process.exit(1);
});

File: examples/python/whale.py
import time
import threading
import logging
import os
import sys
from longportwhale.openapi import Config, HttpClient, TradeContext, PushOrderChanged, TopicType

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger("verify_trade")

# Event to signal completion
completion_event = threading.Event()
order_id = None

def on_order(order_ev: PushOrderChanged):
    global order_id
    logger.info(f"Received order event: {order_ev}")
    if order_id and order_ev.order_id == order_id:
        logger.info("Received expected order event!")
        completion_event.set()

def main():
    global order_id
    
    logger.info("Starting verification...")

    # Load config
    try:
        conf = Config.from_env()
        httpcli = HttpClient.from_env()
    except Exception as e:
        logger.error(f"Failed to load config: {e}")
        sys.exit(1)

    # Initialize TradeContext
    tctx = TradeContext(conf)
    tctx.set_on_order_changed(on_order)
    
    # Subscribe
    logger.info("Subscribing to private events...")
    tctx.subscribe([TopicType.Private])
    time.sleep(2) # Wait for subscription

    # Test HTTP connection
    logger.info("Testing HTTP connection...")
    try:
        asset_req = {
            "account_no": test_account,
            "currency": "HKD"
        }
        asset_res = httpcli.request(method="POST", path="/v1/whaleapi/asset/detail_info", body=asset_req, headers={})
        logger.info(f"Asset query response: {asset_res}")
    except Exception as e:
        logger.error(f"Failed to query asset: {e}")
        # Continue to submit order to see if it persists

    # Submit Order
    test_account = "L6VQEU00121996"
    req = {
        "symbol": "700.HK",
        "order_type": "MO",
        "side": "Buy",
        "submitted_quantity": "100",
        "time_in_force": "Day",
        "account_no": test_account
    }

    logger.info(f"Submitting order: {req}")
    try:
        submit_res = httpcli.request(method="POST", path="/v1/whaleapi/trade/order", body=req, headers={})
        logger.info(f"Submit order response: {submit_res}")
        order_id = submit_res["order_id"]
    except Exception as e:
        logger.error(f"Failed to submit order: {e}")
        sys.exit(1)

    # Wait for event
    logger.info(f"Waiting for order event for order_id: {order_id}...")
    if completion_event.wait(timeout=60):
        logger.info("Verification successful!")
        sys.exit(0)
    else:
        logger.error("Timeout waiting for order event.")
        sys.exit(1)

if __name__ == "__main__":
    main()
